util/network-devp2p/src/node_table.rs

```diff
@@ -344,27 +398,52 @@ impl NodeTable {
 	/// Apply table changes coming from discovery
 	pub fn update(&mut self, mut update: TableUpdates, reserved: &HashSet<NodeId>) {
 		for (_, node) in update.added.drain() {
-			let entry = self.nodes.entry(node.id).or_insert_with(|| Node::new(node.id, node.endpoint.clone()));
-			entry.endpoint = node.endpoint;
-		}
+			let mut add = false;
+			{
+				let entry = self.nodes.entry(node.id).or_insert_with(|| {add = true; Node::new(node.id, node.endpoint.clone())});
+				entry.endpoint = node.endpoint;
+			}
+			if add {
+				if self.ordered_ids.len() == MAX_NODES_IN_TABLE {
+					self.nodes.remove(&self.ordered_ids.pop().expect("ordered_ids is not empty; qed"));
+				};
+				let index = self.get_index_to_insert(None);
+				self.ordered_ids.insert(index, node.id);
+			};
+		};
 		for r in update.removed {
 			if !reserved.contains(&r) {
+				self.ordered_ids.iter().position(|&i| r == i).map(|p| self.ordered_ids.remove(p));
 				self.nodes.remove(&r);
 			}
 		}
 	}
 
 	/// Set last contact as failure for a node
 	pub fn note_failure(&mut self, id: &NodeId) {
+		let mut last_contact = None;
 		if let Some(node) = self.nodes.get_mut(id) {
 			node.last_contact = Some(NodeContact::failure());
+			last_contact = node.last_contact
+		}
+		if let Some(pos) = self.ordered_ids.iter().position(|i| id == i) {
+			self.ordered_ids.remove(pos);
+			let index = self.get_index_to_insert(last_contact);
+			self.ordered_ids.insert(index, *id);
 		}
 	}
 
 	/// Set last contact as success for a node
 	pub fn note_success(&mut self, id: &NodeId) {
+		let mut last_contact = None;
 		if let Some(node) = self.nodes.get_mut(id) {
 			node.last_contact = Some(NodeContact::success());
+			last_contact = node.last_contact
+		}
+		if let Some(pos) = self.ordered_ids.iter().position(|i| id == i) {
```

Comment:
Code duplication with previous method. Please move it to the separate method (like update_ordered_ids(last_contact))